### PR TITLE
[WIP] Refactor test suite paths

### PR DESCRIPTION
**Pull request by jhunkeler**
- Moves nested tests into a single test suite.

- Changes will be the absolute bare minimum required to make things work. Perfection is not the goal right now.

~_Note:_ `stpipe` tests are horrifically broken. At least until I can determine what on Earth it wants from me.~

cc @jdavies-st

### CODE REVIEW

**Comment by jdavies-st** [comment]
Looking good so far!  Of course everything is failing, but that's expected.  I see you did absolute imports by the testing infrastructure.  I.e. in the tests:
```
from jwst.foo import bar
```
instead of
```
from ../../foo import bar
```
This is good!  I recall we had discussed doing this when we did the sprint to move everything to github a couple years ago, but we punted as it was a lot of code to change.

**Comment by stscicrawford** [comment]
For the long term home of the tests, didn't we want to go the other way so that they were in the directories associated with the code that it was testing?  Or am I misunderstanding what is happening here? 

 cc: @pllim

File: setup.cfg [hunk 1]
```diff
@@ -1,7 +1,6 @@
 [metadata]
 package_name = jwst
-description = JWST Pipeline
-author = STScI
+description = JWST Pipeline author = STScI
```
Comment (pllim) [inline setup.cfg]: I gather the deletion of new line was not intentional?

**Comment by pllim** [comment]
I guess there are two separate issues to discuss here:
1. Relative vs absolute import
2. Test dir layout

**Relative vs absolute import**

There was a long discussion over at astropy/astropy#6452 . It seems the consensus was it is fine to use absolute import. I am still not 100% convinced that using absolute import won't get confusing when you test locally on your machine and you have different versions of `jwst` installed, but it is not an issue in Jenkins because only the version being tested will be installed. Perhaps @drdavella can chime in since he brought it up over at Astropy in the first place.

**Test dir layout**

With https://docs.pytest.org/en/latest/goodpractices.html as reference, this PR appears to move the layout from "Tests as part of application code" to "Tests outside application code". This new layout is handy if you don't want to ship the tests with the code to users (i.e., you can prune the whole "tests" subdirectory in the manifest). But if that is not the intention, there is no reason to move them above "jwst" level.

If you ever want to take advantage of the `-P` option like what Astropy test runner uses (e.g., astropy/astropy#7463), keeping tests under each subpackage as-is might be desirable, but it is not the only option you have.

Tests layout that works for you is determined by (but not limited to) the following factors:
* How isolated are the tests? Do they test one and only one step, or multiple at once?
* What tests share which data/config?
* Do they use some common fixtures or utility functions? If so, where are the fixtures/functions located?
* Do you want users (or whoever install it not from source) to be able to run tests?
* Do you want the ability to only run tests specific to a subset of steps?
* Is it more intuitive for you as a team to look for tests at location A versus location B?
* Is the layout easy to maintain going forward? Is it easy to add new tests to?

Hope this helps.

**Comment by sosey** [comment]
@pllim what's the use case for having different versions of jwst installed in the same environment? That sounds like trouble.

**Comment by jhunkeler** [comment]
For the sake showcasing how the project reacts to absolute imports...

## Modify "tests" package to print runtime data

```python
# File: tests/__init__.py
import jwst
import sys


for p in sys.path:
    print('-> {}'.format(p))

print('JWST version: {}'.format(jwst.__version__))
print('JWST origin : {}'.format(jwst.__path__))
```

## Install jwst package into site-packages for effect

```
$ cd jwst
$ python setup.py install
[...]
Installed /Users/jhunk/miniconda3/envs/jwst/lib/python3.6/site-packages/jwst-0.9.4a0.dev49+g83452e54-py3.6-macosx-10.7-x86_64.egg
[...]
Using /Users/jhunk/miniconda3/envs/jwst/lib/python3.6/site-packages
Finished processing dependencies for jwst==0.9.4a0.dev49+g83452e54
```

## Verify the installed jwst package can be resolved

```
$ cd /tmp
$ python -c 'import jwst; print(jwst.__path__)'
['/Users/jhunk/miniconda3/envs/jwst/lib/python3.6/site-packages/jwst-0.9.4a0.dev49+g83452e54-py3.6-macosx-10.7-x86_64.egg/jwst']
```

# Verify how pytest handles this scenario

i.e., `jwst` is installed, however we want to test against the _project_, not `site-packages`.

## Pytest (Inside project root)
```
$ cd jwst
$ pytest -s tests
-> /Users/jhunk/Downloads/jwst
-> /Users/jhunk/miniconda3/envs/jwst/bin
-> /Users/jhunk/miniconda3/envs/jwst/lib/python36.zip
-> /Users/jhunk/miniconda3/envs/jwst/lib/python3.6
-> /Users/jhunk/miniconda3/envs/jwst/lib/python3.6/lib-dynload
-> /Users/jhunk/miniconda3/envs/jwst/lib/python3.6/site-packages
-> /Users/jhunk/miniconda3/envs/jwst/lib/python3.6/site-packages/jwst-0.9.4a0.dev49+g83452e54-py3.6-macosx-10.7-x86_64.egg
-> /Users/jhunk/miniconda3/envs/jwst/lib/python3.6/site-packages/stsci.convolve-2.2.1-py3.6-macosx-10.7-x86_64.egg
-> /Users/jhunk/miniconda3/envs/jwst/lib/python3.6/site-packages/stsci.image-2.3.0-py3.6-macosx-10.7-x86_64.egg
-> /Users/jhunk/miniconda3/envs/jwst/lib/python3.6/site-packages/stsci.imagemanip-1.1.2.dev0-py3.6-macosx-10.7-x86_64.egg
-> /Users/jhunk/miniconda3/envs/jwst/lib/python3.6/site-packages/stsci.imagestats-1.4.1.dev0-py3.6-macosx-10.7-x86_64.egg
-> /Users/jhunk/miniconda3/envs/jwst/lib/python3.6/site-packages/stsci.ndimage-0.10.1.dev0-py3.6-macosx-10.7-x86_64.egg
-> /Users/jhunk/miniconda3/envs/jwst/lib/python3.6/site-packages/stsci.skypac-0.9.10-py3.6.egg
-> /Users/jhunk/miniconda3/envs/jwst/lib/python3.6/site-packages/stsci.stimage-0.2.1.dev0-py3.6-macosx-10.7-x86_64.egg
-> /Users/jhunk/miniconda3/envs/jwst/lib/python3.6/site-packages/stsci.tools-3.4.12.dev32-py3.6.egg
JWST version: 0.9.4a.dev49+g83452e54
JWST origin : ['/Users/jhunk/Downloads/jwst/jwst']
==================================================== test session starts ====================================================
platform darwin -- Python 3.6.5, pytest-3.5.1, py-1.5.3, pluggy-0.6.0
rootdir: /Users/jhunk/Downloads/jwst, inifile: setup.cfg
plugins: xdist-1.22.2, remotedata-0.2.0, forked-0.2
collected 900 items
```

## Pytest (Outside project root)
```
$ cd ..
$ pytest -s jwst/tests
-> /Users/jhunk/Downloads/jwst
-> /Users/jhunk/miniconda3/envs/jwst/bin
-> /Users/jhunk/miniconda3/envs/jwst/lib/python36.zip
-> /Users/jhunk/miniconda3/envs/jwst/lib/python3.6
-> /Users/jhunk/miniconda3/envs/jwst/lib/python3.6/lib-dynload
-> /Users/jhunk/miniconda3/envs/jwst/lib/python3.6/site-packages
-> /Users/jhunk/miniconda3/envs/jwst/lib/python3.6/site-packages/jwst-0.9.4a0.dev49+g83452e54-py3.6-macosx-10.7-x86_64.egg
-> /Users/jhunk/miniconda3/envs/jwst/lib/python3.6/site-packages/stsci.convolve-2.2.1-py3.6-macosx-10.7-x86_64.egg
-> /Users/jhunk/miniconda3/envs/jwst/lib/python3.6/site-packages/stsci.image-2.3.0-py3.6-macosx-10.7-x86_64.egg
-> /Users/jhunk/miniconda3/envs/jwst/lib/python3.6/site-packages/stsci.imagemanip-1.1.2.dev0-py3.6-macosx-10.7-x86_64.egg
-> /Users/jhunk/miniconda3/envs/jwst/lib/python3.6/site-packages/stsci.imagestats-1.4.1.dev0-py3.6-macosx-10.7-x86_64.egg
-> /Users/jhunk/miniconda3/envs/jwst/lib/python3.6/site-packages/stsci.ndimage-0.10.1.dev0-py3.6-macosx-10.7-x86_64.egg
-> /Users/jhunk/miniconda3/envs/jwst/lib/python3.6/site-packages/stsci.skypac-0.9.10-py3.6.egg
-> /Users/jhunk/miniconda3/envs/jwst/lib/python3.6/site-packages/stsci.stimage-0.2.1.dev0-py3.6-macosx-10.7-x86_64.egg
-> /Users/jhunk/miniconda3/envs/jwst/lib/python3.6/site-packages/stsci.tools-3.4.12.dev32-py3.6.egg
JWST version: 0.9.4a.dev49+g83452e54
JWST origin : ['/Users/jhunk/Downloads/jwst/jwst']
==================================================== test session starts ====================================================
platform darwin -- Python 3.6.5, pytest-3.5.1, py-1.5.3, pluggy-0.6.0
rootdir: /Users/jhunk/Downloads/jwst, inifile: setup.cfg
plugins: xdist-1.22.2, remotedata-0.2.0, forked-0.2
collected 900 items
```

## setuptools test command (Inside project root)
```
$ cd jwst
$ python setup.py test --addopts -s
/Users/jhunk/miniconda3/envs/jwst/lib/python3.6/site-packages/setuptools/dist.py:397: UserWarning: Normalizing '0.9.4a.dev49+g83452e54' to '0.9.4a0.dev49+g83452e54'
  normalized_version,
running pytest
running egg_info
writing jwst.egg-info/PKG-INFO
writing dependency_links to jwst.egg-info/dependency_links.txt
writing entry points to jwst.egg-info/entry_points.txt
writing requirements to jwst.egg-info/requires.txt
writing top-level names to jwst.egg-info/top_level.txt
reading manifest file 'jwst.egg-info/SOURCES.txt'
reading manifest template 'MANIFEST.in'
warning: no files found matching '*.csv' under directory 'jwst'
writing manifest file 'jwst.egg-info/SOURCES.txt'
running build_ext
copying build/lib.macosx-10.7-x86_64-3.6/jwst/tweakreg/chelp.cpython-36m-darwin.so -> jwst/tweakreg
-> /Users/jhunk/Downloads/jwst
-> /Users/jhunk/Downloads/jwst/.eggs/pytest_runner-4.2-py3.6.egg
-> /Users/jhunk/Downloads/jwst
-> relic
-> /Users/jhunk/miniconda3/envs/jwst/lib/python36.zip
-> /Users/jhunk/miniconda3/envs/jwst/lib/python3.6
-> /Users/jhunk/miniconda3/envs/jwst/lib/python3.6/lib-dynload
-> /Users/jhunk/miniconda3/envs/jwst/lib/python3.6/site-packages
-> /Users/jhunk/miniconda3/envs/jwst/lib/python3.6/site-packages/jwst-0.9.4a0.dev49+g83452e54-py3.6-macosx-10.7-x86_64.egg
-> /Users/jhunk/miniconda3/envs/jwst/lib/python3.6/site-packages/stsci.convolve-2.2.1-py3.6-macosx-10.7-x86_64.egg
-> /Users/jhunk/miniconda3/envs/jwst/lib/python3.6/site-packages/stsci.image-2.3.0-py3.6-macosx-10.7-x86_64.egg
-> /Users/jhunk/miniconda3/envs/jwst/lib/python3.6/site-packages/stsci.imagemanip-1.1.2.dev0-py3.6-macosx-10.7-x86_64.egg
-> /Users/jhunk/miniconda3/envs/jwst/lib/python3.6/site-packages/stsci.imagestats-1.4.1.dev0-py3.6-macosx-10.7-x86_64.egg
-> /Users/jhunk/miniconda3/envs/jwst/lib/python3.6/site-packages/stsci.ndimage-0.10.1.dev0-py3.6-macosx-10.7-x86_64.egg
-> /Users/jhunk/miniconda3/envs/jwst/lib/python3.6/site-packages/stsci.skypac-0.9.10-py3.6.egg
-> /Users/jhunk/miniconda3/envs/jwst/lib/python3.6/site-packages/stsci.stimage-0.2.1.dev0-py3.6-macosx-10.7-x86_64.egg
-> /Users/jhunk/miniconda3/envs/jwst/lib/python3.6/site-packages/stsci.tools-3.4.12.dev32-py3.6.egg
JWST version: 0.9.4a.dev49+g83452e54
JWST origin : ['/Users/jhunk/Downloads/jwst/jwst']
==================================================== test session starts ====================================================
platform darwin -- Python 3.6.5, pytest-3.5.1, py-1.5.3, pluggy-0.6.0
rootdir: /Users/jhunk/Downloads/jwst, inifile: setup.cfg
plugins: xdist-1.22.2, remotedata-0.2.0, forked-0.2
collected 900 items
```

**Comment by pllim** [comment]
Thanks, Joe! I think you have convinced me that absolute import is fine. :smile:

**Comment by stscicrawford** [comment]
I think there has been a pretty substantial refactor of the test suite since this time, and so I am going to close this.   For now, we still want to keep the unit tests with the code, but I'll open an issue as I think we do want to move to absolute imports.  

Feel free to reopen if appropriate.

**Comment by pllim** [comment]
Re: absolute imports -- there is a precedence in astropy/astropy#8200 , so definitely worth revisiting FWIW

**Comment by stscicrawford** [comment]
Opened issued #2991 to track it.

**Pull request closed without merging**